server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -210,71 +212,192 @@ public static KeyStoreWrapper load(Path configDir) throws IOException {
+        try (ByteArrayInputStream bytesStream = new ByteArrayInputStream(dataBytes);
+             DataInputStream input = new DataInputStream(bytesStream)) {
+            int saltLen = input.readInt();
+            salt = new byte[saltLen];
+            if (input.read(salt) != saltLen) {
+                throw new SecurityException("Keystore has been corrupted or tampered with");
+            }
+            int ivLen = input.readInt();
+            iv = new byte[ivLen];
+            if (input.read(iv) != ivLen) {
+                throw new SecurityException("Keystore has been corrupted or tampered with");
+            }
+            int encryptedLen = input.readInt();
+            encryptedBytes = new byte[encryptedLen];
+            if (input.read(encryptedBytes) != encryptedLen) {
+                throw new SecurityException("Keystore has been corrupted or tampered with");
+            }
+        }
+
+        Cipher cipher = createCipher(Cipher.DECRYPT_MODE, password, salt, iv);
+        try (ByteArrayInputStream bytesStream = new ByteArrayInputStream(encryptedBytes);
+             CipherInputStream cipherStream = new CipherInputStream(bytesStream, cipher);
+             DataInputStream input = new DataInputStream(cipherStream)) {
+
+            entries.set(new HashMap<>());
+            int numEntries = input.readInt();
+            while (numEntries-- > 0) {
+                String setting = input.readUTF();
+                EntryType entryType = EntryType.valueOf(input.readUTF());
+                int entrySize = input.readInt();
+                byte[] entryBytes = new byte[entrySize];
+                if (input.read(entryBytes) != entrySize) {
+                    throw new SecurityException("Keystore has been corrupted or tampered with");
+                }
+                entries.get().put(setting, new Entry(entryType, entryBytes));
+            }
+        }
+    }
+
+    /** Encrypt the keystore entries and return the encrypted data. */
+    private byte[] encrypt(char[] password, byte[] salt, byte[] iv) throws GeneralSecurityException, IOException {
+        assert isLoaded();
+
+        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+        Cipher cipher = createCipher(Cipher.ENCRYPT_MODE, password, salt, iv);
+        try (CipherOutputStream cipherStream = new CipherOutputStream(bytes, cipher);
+             DataOutputStream output = new DataOutputStream(cipherStream)) {
+
+            output.writeInt(entries.get().size());
+            for (Map.Entry<String, Entry> mapEntry : entries.get().entrySet()) {
+                output.writeUTF(mapEntry.getKey());
+                Entry entry = mapEntry.getValue();
+                output.writeUTF(entry.type.name());
+                output.writeInt(entry.bytes.length);
+                output.write(entry.bytes);
+            }
+        }
+
+        return bytes.toByteArray();
+    }
+
+    private void decryptLegacyEntries() throws GeneralSecurityException, IOException {
```

Comment:
This won't work if the JVM is in FIPS approved mode only as `PBE` is not available and ` SecretKeyFactory.getInstance("PBE");` will throw an Exception. 
I'm thinking of the case where someone wants to upgrade to 6.3.0 and at the same time enable FIPS approved only mode. We can highlight this and document manual steps that will ensure the keystore gets upgraded before FIPS approved mode is enabled.

Reply:
We can add an assumeFalse once we are ready to test that.